V1.0.0 RC (#182)
OWASP/maswe: weaknesses/MASVS-RESILIENCE/MASWE-0059.md

Author: Sergio: DONE

---
title: Device Attestation Not Implemented
id: MASWE-0059
alias: device-attestation
requirement: "The app implements device attestation."
platform: [android, ios]
profiles: [R]
threat: MAS-THREAT-0059
attacks: [MAS-ATTACK-0065, MAS-ATTACK-0066, MAS-ATTACK-0068]
mappings:
  masvs-v1: [MSTG-RESILIENCE-10]
  masvs-v2: [MASVS-RESILIENCE-1]

>   cwe: [693]

Reviewer: ```suggestion
```

  maswe-beta: [MASWE-0100]
refs:
- https://developer.android.com/google/play/integrity
- https://support.google.com/googleplay/android-developer/answer/11395166?hl=en
- https://www.youtube.com/watch?v=TyxL78e5Bag
- https://github.com/1nikolas/play-integrity-checker-app
- https://developer.apple.com/videos/play/wwdc2021/10244/
- https://developer.apple.com/documentation/devicecheck/preparing-to-use-the-app-attest-service
- https://github.com/iansampson/AppAttest
- https://github.com/firebase/firebase-ios-sdk/blob/v8.15.0/FirebaseAppCheck/Sources/AppAttestProvider/DCAppAttestService%2BFIRAppAttestService.h

> - https://blog.restlesslabs.com/john/ios-app-attest

Reviewer: ```suggestion
- https://developer.android.com/google/play/integrity
- https://developer.android.com/google/play/integrity/verdicts
- https://developer.android.com/google/play/integrity/standard
- https://developer.android.com/google/play/integrity/classic
- https://developer.apple.com/documentation/devicecheck
- https://developer.apple.com/documentation/devicecheck/accessing-and-modifying-per-device-data
- https://developer.apple.com/documentation/devicecheck/establishing-your-app-s-integrity
- https://developer.apple.com/documentation/devicecheck/validating-apps-that-connect-to-your-server
- https://github.com/Oliver-Binns/app-attest
- https://github.com/VisionR1/KeyAttestation
- https://github.com/JingMatrix/TEESimulator
- https://github.com/5ec1cff/TrickyStore
```

status: new
---

## Overview

> This weakness occurs when an app does not implement device attestation, so its backend cannot distinguish requests from genuine, uncompromised devices from those coming from rooted, emulated, tampered, or automated environments.

Reviewer: ```suggestion
This weakness occurs when an app does not implement device attestation, so its backend cannot distinguish requests made from genuine, uncompromised devices from those coming from rooted, emulated, tampered, or automated environments.
```

> Device attestation uses platform services, such as the Android Play Integrity API or iOS DeviceCheck and App Attest, to give the backend cryptographic assurance about the integrity of the device and platform the app runs on. Without it, the backend must trust whatever the client claims. Attestation results must be verified server-side, including nonce-based freshness, to be meaningful; a verdict checked only on the client is just another bypassable local check.

Reviewer: It's a bit tricky because unlike PlayIntegrity, Apple's solution does not check if the device is compromised (e.g. jailbroken). Its main purposes are attesting that the requesting app has not been repackaged/resigned (AppAttest) and providing a device token + a remote saved state to said token (devicecheck).

However, using this framework may work as mitigation since it requires the device having a working SecureEnclave which might not be the case for some emulator implementations.

```suggestion
Device attestation uses platform services, such as the Android Play Integrity API or iOS DeviceCheck and App Attest, to provide the backend with cryptographically verifiable evidence about the device associated with a request. The evidence and security properties differ by platform and service. Without device attestation, the backend cannot independently verify device-origin claims made by the client. The backend must validate attestation evidence and apply the service-specific request-binding, freshness, and replay protections; client-side evaluation is another bypassable local check.

On Android, Play Integrity can provide server-verifiable assurance that the device has not been compromised in ways covered by its device-integrity verdicts, such as an unlocked bootloader or an unrecognized operating-system image. On iOS, DeviceCheck and App Attest do not provide equivalent assurance about operating-system compromise. They can establish that evidence originates from genuine Apple hardware; App Attest also binds that evidence to a legitimate app instance, but neither service attests the integrity of iOS.
```

## Modes of Introduction

- **No Attestation Integrated**: Not using the platform's attestation services at all, leaving the backend with no device-integrity signal.
- **Client-Side-Only Verification**: Requesting attestation but evaluating the verdict in the app instead of verifying it server-side.
- **Missing Freshness Guarantees**: Verifying attestation without a server-issued nonce or timeliness check, allowing verdicts to be replayed.

> - **Verdicts Not Enforced**: Collecting attestation results but not gating sensitive operations on them.

Reviewer: The developer could not be verifying the attestation report in the backend

```suggestion
- **Verdicts Not Enforced**: Collecting attestation results but not gating sensitive operations on them.
- **Incomplete Evidence Validation**: Accepting device-attestation evidence without applying the service-specific checks for request binding, replay protection, and the device claims required by the backend's policy.
```

## Impact

> - **Compromise of System Integrity and Business Operations**: Attackers can drive the backend with automated or tampered clients, resulting in fraud, scraping, fake accounts, and abuse of the app owner's services.

Reviewer: ```suggestion
- **Compromise of System Integrity and Business Operations**: Attackers can drive the backend with automated or tampered clients, resulting in fraud, scraping, fake accounts, and abuse of the app owner's services.
- **Bypass of Protection Mechanisms**: Attackers can leverage tampered running environments to bypass the app's security checks or feature restrictions, resulting in unauthorized access to protected app functionality.
```

---
title: Device Attestation Not Implemented
id: MASWE-0059
alias: device-attestation
requirement: "The app implements device attestation."
platform: [android, ios]
profiles: [R]
threat: MAS-THREAT-0059
attacks: [MAS-ATTACK-0065, MAS-ATTACK-0066, MAS-ATTACK-0068]
mappings:
  masvs-v1: [MSTG-RESILIENCE-10]
  masvs-v2: [MASVS-RESILIENCE-1]
  maswe-beta: [MASWE-0100]
refs:
- https://developer.android.com/google/play/integrity
- https://developer.android.com/google/play/integrity/verdicts
- https://developer.android.com/google/play/integrity/standard
- https://developer.android.com/google/play/integrity/classic

> - https://developer.apple.com/documentation/devicecheck

Reviewer: Adding references for standard key attestation

```suggestion
- https://developer.android.com/google/play/integrity/classic
- https://github.com/android/keyattestation
- https://source.android.com/docs/security/features/keystore/attestation
- https://grapheneos.org/articles/attestation-compatibility-guide
- https://developer.apple.com/documentation/devicecheck
```

- https://developer.apple.com/documentation/devicecheck/accessing-and-modifying-per-device-data
- https://developer.apple.com/documentation/devicecheck/establishing-your-app-s-integrity
- https://developer.apple.com/documentation/devicecheck/validating-apps-that-connect-to-your-server
- https://github.com/Oliver-Binns/app-attest
- https://github.com/VisionR1/KeyAttestation
- https://github.com/JingMatrix/TEESimulator
- https://github.com/5ec1cff/TrickyStore
status: new
---

## Overview

This weakness occurs when an app does not implement device attestation, so its backend cannot distinguish requests made from genuine, uncompromised devices from those coming from rooted, emulated, tampered, or automated environments.

> Device attestation uses platform services, such as the Android Play Integrity API or iOS DeviceCheck and App Attest, to provide the backend with cryptographically verifiable evidence about the device associated with a request. The evidence and security properties differ by platform and service. Without device attestation, the backend cannot independently verify device-origin claims made by the client. The backend must validate attestation evidence and apply the service-specific request-binding, freshness, and replay protections; client-side evaluation is another bypassable local check.

Reviewer: Mention hardware attestation API

```suggestion
Device attestation uses platform services, such as the standard Android hardware attestation API, the Play Integrity API or iOS DeviceCheck and App Attest, to provide the backend with cryptographically verifiable evidence about the device associated with a request. The evidence and security properties differ by platform and service. Without device attestation, the backend cannot independently verify device-origin claims made by the client. The backend must validate attestation evidence and apply the service-specific request-binding, freshness, and replay protections; client-side evaluation is another bypassable local check.
```

> On Android, Play Integrity can provide server-verifiable assurance that the device has not been compromised in ways covered by its device-integrity verdicts, such as an unlocked bootloader or an unrecognized operating-system image. On iOS, DeviceCheck and App Attest do not provide equivalent assurance about operating-system compromise. They can establish that evidence originates from genuine Apple hardware; App Attest also binds that evidence to a legitimate app instance, but neither service attests the integrity of iOS.

Reviewer 2: @cpholguera @sgIOlas 

I know the MASWE is already reviewed, but during the review of the referenced MASWE-0059 I noticed that the method of validating the key attestation on Android (namely, [RootOfTrust](https://source.android.com/docs/security/features/keystore/attestation#rootoftrust-fields)) was not mentioned. 

This can be an alternative to Play Integrity. It does not always equal "rooted" but the developer can use it as a signal on the server side to detect current or previous attempts to root the device.

Reviewer 3: RootOfTrust would also be device specific already? I can imagine a device with a different attestation API generating different signed objects? Or maybe this is low-level enough to be generic ?

Reviewer 2: @TheDauntless There can be differences between the different vendors / platform versions. 

All the attestation relevant information are stored in the x509 certificate `attestation extension` which has the OID `1.3.6.1.4.1.11129.2.1.17`. 

There are different versions (https://source.android.com/docs/security/features/keystore/attestation#schema): 

<img width="907" height="451" alt="Screenshot 2026-07-31 at 12 42 24" src="https://github.com/user-attachments/assets/104c73f6-aaac-4739-b9dc-8a6e359cbf50" />

All versions support `RootOfTrust` but only versions since 3 contain the properly `verifiedBootHash`:

Version 1: 

```
RootOfTrust ::= SEQUENCE {
    verifiedBootKey            OCTET_STRING,
    deviceLocked               BOOLEAN,
    verifiedBootState          VerifiedBootState,
}
```

Current version 500:

```
RootOfTrust ::= SEQUENCE {
    verifiedBootKey            OCTET_STRING,
    deviceLocked               BOOLEAN,
    verifiedBootState          VerifiedBootState,
    verifiedBootHash           OCTET_STRING,
}
```

In the end, the schema depends on the Android version and the hardware support. 

So for the app developer the **API is the same**. If the device supports TEE / key attestation, the key attestation certificate will always be accessible through the same API, but the schema of the attestation extension can change. This has to be considered when decoding it.

Author: Captured in new issue

> On Android, Play Integrity can provide server-verifiable assurance that the device has not been compromised in ways covered by its device-integrity verdicts, such as an unlocked bootloader or an unrecognized operating-system image. On iOS, DeviceCheck and App Attest do not provide equivalent assurance about operating-system compromise. They can establish that evidence originates from genuine Apple hardware; App Attest also binds that evidence to a legitimate app instance, but neither service attests the integrity of iOS.

Reviewer: Mention hardware attestation API

```suggestion
On Android, both the hardware-backed attestation and Play Integrity APIs can provide server-verifiable assurance that the device has not been compromised in ways covered by its device-integrity verdicts or signals, such as an unlocked bootloader, an unrecognized operating-system image or an outdated security patch level. Their main differences include:

- **Google Mobile Services (GMS) Requirement**: Direct hardware-backed key attestation uses Android Keystore and does not require Google Play services on the device to request an attestation. Play Integrity depends on Google Play components and Google services.
- **Verdict Generation and Policy Enforcement**: With direct key attestation, the backend receives signed properties and defines which boot states, operating-system signing keys, patch levels, and application identities it accepts. Play Integrity evaluates the available signals and returns Google-defined verdict labels. In both cases, the developer decides how the backend responds.

On iOS, DeviceCheck and App Attest do not provide equivalent assurance about operating-system compromise. They can establish that evidence originates from genuine Apple hardware; App Attest also binds that evidence to a legitimate app instance, but neither service attests the integrity of iOS.
```

- **Missing Freshness Guarantees**: Verifying attestation without a server-issued nonce or timeliness check, allowing verdicts to be replayed.
- **Verdicts Not Enforced**: Collecting attestation results but not gating sensitive operations on them.
- **Incomplete Evidence Validation**: Accepting device-attestation evidence without applying the service-specific checks for request binding, replay protection, and the device claims required by the backend's policy.

## Impact

- **Compromise of System Integrity and Business Operations**: Attackers can drive the backend with automated or tampered clients, resulting in fraud, scraping, fake accounts, and abuse of the app owner's services.
- **Bypass of Protection Mechanisms**: Attackers can leverage tampered running environments to bypass the app's security checks or feature restrictions, resulting in unauthorized access to protected app functionality.
- **Financial Loss**: Attackers can abuse promotions, premium features, or transaction flows from unattested environments, resulting in direct monetary loss to the app owner.

## Mitigations

> - **Integrate Platform Attestation**: Use the Android Play Integrity API and iOS DeviceCheck / App Attest to obtain device- and app-integrity verdicts.

Reviewer 2: ```suggestion
- **Integrate Platform Attestation**: Validate the [rootOfTrust](https://source.android.com/docs/security/features/keystore/attestation#rootoftrust-fields) fields from the key attestation extension or use Play Integrity API and iOS DeviceCheck / App Attest to obtain device- and app-integrity verdicts.
```
